extensions/git/src/git.ts

```diff
@@ -138,11 +138,30 @@ function findGitWin32(onLookup: (path: string) => void): Promise<IGit> {
 		.then(undefined, () => findGitWin32InPath(onLookup));
 }
 
-export function findGit(hint: string | undefined, onLookup: (path: string) => void): Promise<IGit> {
-	const first = hint ? findSpecificGit(hint, onLookup) : Promise.reject<IGit>(null);
+export function findGit(hints: string | string[] | undefined, onLookup: (path: string) => void): Promise<IGit> {
```

Comment:
I feel like this function could use a little decomposition. Removing nested if/else blocks, adding async/await, and adding some helpers would make this a lot cleaner.  Something like

```
function findPlatformGit(platform: string, onLookup: (path: string) => void): Promise<IGit> {
    switch (platform) {
        case 'darwin': return findGitDarwin(onLookup);
        case 'win32': return findGitWin32(onLookup);
        default: return findSpecificGit('git', onLookup);
    }
}

export async function findGit(hints: string | string[] | undefined, onLookup: (path: string) => void): Promise<IGit> {
    const createHintsArray = (hint: string | string[] | undefined, fallback: Promise<IGit>) => {
        if (!hint) {
            return [fallback];
        }
        if (Array.isArray(hint)) {
            return hint.map(h => findSpecificGit(h, onLookup));
        }
        return [findSpecificGit(h, onLookup)];
    }
    const potentialGitPaths = createHintsArray(hints, findPlatformGit(process.platform, onLookup));

    for (const potentialPath of potentialGitPaths) {
        try {
            return await potentialPath;
        } catch (e) { /** noop */ }
    }

    return Promise.reject(new Error('Git installation not found.'));
}
```